lib/saml/user.rb

```diff
@@ -17,35 +17,37 @@ def last_signed_in
       Time.current.utc
     end
 
-    def loa
-      { current: loa_current, highest: loa_highest(attributes) }
-    end
-
     def to_hash
       Hash[serializable_attributes.map { |k| [k, @decorated.send(k)] }]
     end
 
+    def authn_context
+      return 'dslogon' if dslogon?
+      return 'mhv' if mhv?
+      nil
+    end
+
     private
 
     def serializable_attributes
-      @decorated.send(:serializable_attributes) + %i(last_signed_in loa)
+      @decorated.send(:serializable_attributes) + %i(authn_context last_signed_in)
     end
 
-    def decorator_constant
-      "SAML::UserAttributes::#{@authn_context.upcase}".safe_constantize ||
-        SAML::UserAttributes::IdMe
+    def dslogon?
+      attributes.to_h.keys.include?('dslogon_uuid')
     end
 
-    def loa_current
-      @raw_loa ||= REXML::XPath.first(@saml_response.decrypted_document, '//saml:AuthnContextClassRef')&.text
-      LOA::MAPPING[@raw_loa]
+    def mhv?
+      attributes.to_h.keys.include?('mhv_uuid')
     end
 
-    def loa_highest(attributes)
-      Rails.logger.warn 'LOA.highest is nil!' if (loa = attributes['level_of_assurance']&.to_i).nil?
-      loa_highest = loa || loa_current
-      Rails.logger.warn 'LOA.highest is less than LOA.current' if loa_highest < loa_current
-      [loa_current, loa_highest].max
+    def decorator_constant
+      case authn_context
+      when 'mhv'; then 'SAML::UserAttributes::MHV'.safe_constantize
```

Comment:
may be missing something but what's `safe_constantize` used for here? could you just have the namespace as on the case's else?

Reply:
haha, yes. I'll change that. Legacy artifact of having tried to build the class name based off of upcasing the authn_context.